blogs/2021/09/06/bracket-pair-colorization.md

```diff
@@ -0,0 +1,311 @@
+It's an awesome extension with over 3 million installs (plus 6 million installs of its [predecessor](https://marketplace.visualstudio.com/items?itemName=CoenraadS.bracket-pair-colorizer)) and a prime example of VS Code's customizability.
+
+<img src="./on-off-comparison.drawio.svg" alt="Comparing Colorization On vs. Off" width="1000"/>
+
+Unfortunately, it has a serious performance problem: When inserting a single bracket at the beginning of the infamous [checker.ts](https://github.com/microsoft/TypeScript/blob/8362a0f929d74ff46828016ec67c05744a8dbb3c/src/compiler/checker.ts) file, it takes about 10 seconds until the colors of the bracket pairs update.
+During these 10 seconds of processing the text change, the extension host becomes unresponsive and all features that are powered by extensions, such as auto-completion or diagnostics, stop functioning.
+
+<img src="./checker_ts-extension.gif" alt="Extension needs more than 10 seconds to process text changes in checker.ts" width="500"/>
+
+We reimplemented the extension in the core of VS Code and brought this time down to less than a millisecond.
+
+By leveraging (2,3)-trees, recursion-free tree-traversal, bit-arithmetic, incremental parsing and other techniques,
+we reduced the update complexity from $\mathcal{O}(E + N)$ to $\mathcal{O}(E + \mathrm{log}^3 N)$ with $N$ being the document size and $E$ the edit size, assuming the nesting level of bracket pairs is bounded by $\mathcal{O}(\mathrm{log} N)$.
+By reusing the existing tokens from the renderer and its incremental token update mechanism, we gain another massive (constant) speedup. Updates are no longer noticable, even for files with hundreds of thousands of bracket pairs:
+
+<img src="./checker_ts-native.gif" alt="Native implementation needs less than a millisecond to process text changes in checker.ts" width="500"/>
+
+## The Challenge Of Bracket Pair Colorization
+
+Bracket pair colorization is all about quickly computing all brackets and their nesting level in the viewport (i.e. a given range in the document).
+Unfortunately, the nesting level of a bracket depends on all characters preceding it: Replacing any character with the open bracket `{` usually increases the nesting level of all following brackets.
+Thus, a lower bound on the time complexity of querying the nesting levels of brackets at the end of a document of size $N$ is $\mathcal{O}(N)$: We have to look at every character at least once.
+
+![Changing a single character influences the nesting level of all subsequent brackets](./level-depends-on-all-previous-characters.dio.svg)
+
+The bracket pair colorizer extension processes the entire document again whenever a bracket is inserted or removed and then sends the location of all bracket pairs to the renderer.
```

Comment:
I made it more clear that the extension does the best it can do.
Not pointing out it's linear update complexity is counterproductive when we talk about our speedup though. After all, if the extension was fast enough, we would not have reimplemented it.